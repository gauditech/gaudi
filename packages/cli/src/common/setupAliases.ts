import path from "path";

import { addAliases } from "module-alias";

/**
 * NOTE: Make sure to keep aliases in sync with `tsconfig.json` and `jest.config.js`.
 */

addAliases({
  "@cli": path.join(__dirname, "src"),
});
