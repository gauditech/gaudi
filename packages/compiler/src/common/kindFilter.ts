type Kind = string | number | symbol | boolean | undefined | null;

export type FilteredByKind<T extends { kind: Kind }, kind = T["kind"]> = kind extends T["kind"]
  ? Exclude<{ [K in keyof T]: K extends "kind" ? kind : T[K] }, { kind: never }>
  : never;

export type RejectedByKind<T extends { kind: Kind }, kind = T["kind"]> = Exclude<
  FilteredByKind<T>,
  { kind: kind }
>;

export function kindFilter<i extends { kind: Kind }[], const k extends i[number]["kind"]>(
  input: i,
  ...kinds: k[]
): FilteredByKind<i[number], k>[] {
  return input.filter((i) => {
    for (const kind of kinds) {
      if (kind === i.kind) return true;
    }
    return false;
  }) as FilteredByKind<i[number], k>[];
}

export function kindReject<i extends { kind: Kind }[], const k extends i[number]["kind"]>(
  input: i,
  ...kinds: k[]
): RejectedByKind<i[number], k>[] {
  return input.filter((i) => {
    for (const kind of kinds) {
      if (kind !== i.kind) return true;
    }
    return false;
  }) as RejectedByKind<i[number], k>[];
}

export function kindFind<i extends { kind: Kind }[], const k extends i[number]["kind"]>(
  input: i,
  ...kinds: k[]
): FilteredByKind<i[number], k> | undefined {
  return input.find((i) => {
    for (const kind of kinds) {
      if (kind === i.kind) return true;
    }
    return false;
  }) as FilteredByKind<i[number], k> | undefined;
}
