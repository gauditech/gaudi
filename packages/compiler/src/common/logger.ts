import { AsyncLocalStorage } from "node:async_hooks";
import util from "util";

import chalk from "chalk";
import _ from "lodash";
import { createLogger, format, transports } from "winston";

const als = new AsyncLocalStorage<string[]>();

/**
 * Logger implementation
 */

const winstonLogger = createLogger({
  level: "silly",
  format: format.combine(
    format.timestamp({
      format: "YYYY-MM-DD HH:mm:ss",
    }),
    format.label(),
    format.errors({ stack: true }),
    format.splat(),
    format.align(),
    format.padLevels(),
    format.printf(({ level, label, message, timestamp, stack }) => {
      const cStack = coloredStack(stack);
      const cTimestamp = coloredTs(timestamp);
      const cLevel = coloredLevel(level);
      const cLabel = coloredLabel(label);
      const msg = stack ? "" : padMessage(message);

      return `${cTimestamp} ${cLevel}${cLabel} ${msg}${cStack}`;
    })
  ),
  transports: [new transports.Console({})],
});

type LogLevel = "error" | "warn" | "info" | "verbose" | "debug" | "silly";

interface LogConfig {
  [category: string]: LogLevel;
}

const LOG_CATEGORIES: LogConfig = {
  "": "silly",
  sql: "debug",
  http: "debug",
  debug: "silly",
};

type Loggable = unknown;

export class Logger {
  private _category: string | null = null;
  private _forceLevel: LogLevel | undefined;

  constructor(category?: string, level?: LogLevel) {
    this._category = category ?? null;
    this._forceLevel = level;
  }

  silly(label: Loggable, message: Loggable): void;
  silly(message: Loggable): void;
  silly(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("silly", args[1], args[0]);
    } else {
      // only message is passed
      this._log("silly", args[0], undefined);
    }
  }

  debug(label: Loggable, message: Loggable): void;
  debug(message: Loggable): void;
  debug(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("debug", args[1], args[0]);
    } else {
      // only message is passed
      this._log("debug", args[0], undefined);
    }
  }

  info(label: Loggable, message: Loggable): void;
  info(message: Loggable): void;
  info(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("info", args[1], args[0]);
    } else {
      // only message is passed
      this._log("info", args[0], undefined);
    }
  }

  error(label: Loggable, message: Loggable): void;
  error(message: Loggable): void;
  error(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("error", args[1], args[0]);
    } else {
      // only message is passed
      this._log("error", args[0], undefined);
    }
  }

  wrap<X>(name: string, cb: () => X): X {
    // NOTE: We don't currently implement anything that can utilize `als`
    const ctx = als.getStore();
    const newCtx = [...(ctx ?? []), name];
    return als.run(newCtx, cb) as X;
  }

  static specific(name: string, level?: LogLevel) {
    return new Logger(name, level);
  }

  private _log(level: LogLevel, message: Loggable, label: Loggable | undefined) {
    const minLogLevel = this._forceLevel ?? LOG_CATEGORIES[this._category || ""] ?? "silly";
    if (Logger.compareLevels(level, minLogLevel)) {
      this._do_log(level, message, label);
    }
  }

  protected _do_log(level: LogLevel, message: Loggable, label: Loggable | undefined): void {
    const msgObj =
      message instanceof Error || typeof message === "string"
        ? message
        : util.inspect(message, { depth: 10, colors: true });
    winstonLogger.log(level, msgObj as any, { label });
  }

  private static compareLevels(a: LogLevel, b: LogLevel): boolean {
    const levels: LogLevel[] = ["error", "warn", "info", "verbose", "debug", "silly"];
    return levels.indexOf(a) - levels.indexOf(b) <= 0;
  }
}

const logger = new Logger();
export default logger;

/**
 * Helper functions.
 */

function coloredLevel(level: string): string {
  const l = chalk.bold(chalk.white(` ${level.toUpperCase().padEnd(5)} `));
  switch (level) {
    case "silly": {
      return chalk.bgYellow(l);
    }
    case "debug": {
      return chalk.bgBlue(l);
    }
    case "info": {
      return chalk.bgGreen(l);
    }
    case "error": {
      return chalk.bgRed(l);
    }
    default: {
      return level;
    }
  }
}

function coloredTs(ts: string): string {
  return chalk.bold(`[${ts}]`);
}

function coloredStack(stack: string): string {
  if (!stack) {
    return "";
  }
  const [err, ...s] = stack.split("\n").map((line) => `${_.repeat("\t", 2)}${line}`);
  return `${chalk.bold(chalk.red(err))}\n${chalk.gray(s.join("\n"))}`;
}

function coloredLabel(label: string | undefined): string {
  return label ? chalk.bgGray(chalk.yellowBright(chalk.bold(`[${label}]`))) : "";
}

function padMessage(msg: string): string {
  return msg
    .split("\n")
    .map((line, index) => (index > 0 ? `\t\t${line}` : line))
    .join("\n");
}
