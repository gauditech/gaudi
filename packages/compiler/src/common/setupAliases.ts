import path from "path";

import { addAliases } from "module-alias";

/**
 * NOTE: Make sure to keep aliases in sync with `tsconfig.json` and `jest.config.js`.
 */

addAliases({
  "@compiler": path.join(__dirname, ".."),
});
