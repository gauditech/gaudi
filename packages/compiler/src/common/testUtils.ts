import { compileToAST } from "@compiler/compiler";
import { compilerErrorsToString } from "@compiler/compiler/compilerError";
import { migrate } from "@compiler/compiler/migrate";
import { compose } from "@compiler/composer/composer";
import { Definition } from "@compiler/types/definition";

/**
 * Helper function that compiles directly to definition. This is used in tests.
 */
export function compileFromString(source: string): Definition {
  const inputs = [{ source }];
  const { ast, errors } = compileToAST([{ source }]);
  if (errors) {
    const errorString = compilerErrorsToString(inputs, errors);
    throw new Error(`Failed to compile:\n${errorString}`);
  }
  return compose(migrate(ast));
}
