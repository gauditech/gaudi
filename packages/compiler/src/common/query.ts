import { assertUnreachable, ensureEqual } from "@compiler/common/utils";
import { SelectDef, SelectItem, TypedExprDef } from "@compiler/types/definition";
import _ from "lodash";
import { match } from "ts-pattern";

export function transformSelectPath(select: SelectDef, from: string[], to: string[]): SelectDef {
  return select.map((item: SelectItem): SelectItem => {
    return match(item)
      .with({ kind: "expression" }, (item) => ({
        ...item,
        expr: transformExpressionPaths(item.expr, from, to),
      }))
      .with({ kind: "model-hook" }, (item) => ({
        ...item,
        namePath: transformNamePath(item.namePath, from, to),
      }))
      .with({ kind: "nested-select" }, (item) => {
        return {
          ...item,
          namePath: transformNamePath(item.namePath, from, to),
          select: transformSelectPath(item.select, from, to),
        };
      })
      .exhaustive();
  });
}

export function transformNamePath(path: string[], from: string[], to: string[]): string[] {
  ensureEqual(
    _.isEqual(from, _.take(path, from.length)),
    true,
    `Cannot transform name path: ${path.join(".")} doesn't start with ${from.join(".")}`
  );
  return [...to, ..._.drop(path, from.length)];
}

export function transformNamePaths(paths: string[][], from: string[], to: string[]): string[][] {
  return paths.map((path) => transformNamePath(path, from, to));
}

export function transformExpressionPaths(
  exp: TypedExprDef,
  from: string[],
  to: string[]
): TypedExprDef {
  if (exp === undefined) {
    return undefined;
  }
  switch (exp.kind) {
    case "literal":
    case "variable": {
      return exp;
    }
    case "alias": {
      return { ...exp, namePath: transformNamePath(exp.namePath, from, to) };
    }
    case "function": {
      return {
        ...exp,
        args: exp.args.map((arg) => transformExpressionPaths(arg, from, to)),
      };
    }
    case "aggregate-function": {
      return { ...exp, sourcePath: transformNamePath(exp.sourcePath, from, to) };
    }
    case "in-subquery": {
      return {
        ...exp,
        sourcePath: transformNamePath(exp.sourcePath, from, to),
        lookupExpression: transformExpressionPaths(exp, from, to),
      };
    }
    case "array": {
      return {
        ...exp,
        elements: exp.elements.map((arg) => transformExpressionPaths(arg, from, to)),
      };
    }
    default: {
      assertUnreachable(exp);
    }
  }
}
