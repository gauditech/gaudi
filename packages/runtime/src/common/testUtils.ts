import { compileToAST } from "@gaudi/compiler/dist/compiler";
import { compilerErrorsToString } from "@gaudi/compiler/dist/compiler/compilerError";
import { migrate } from "@gaudi/compiler/dist/compiler/migrate";
import { compose } from "@gaudi/compiler/dist/composer/composer";
import {
  CustomManyEndpointDef,
  Definition,
  QueryAction,
  QueryDef,
} from "@gaudi/compiler/dist/types/definition";

import { QueryTree } from "@runtime/query/build";
import { NestedRow, QueryExecutor } from "@runtime/query/exec";
import { Storage } from "@runtime/server/context";

/**
 * Helper function that compiles directly to definition. This is used in tests.
 */
export function compileFromString(source: string): Definition {
  const inputs = [{ source }];
  const { ast, errors } = compileToAST([{ source }]);
  if (errors) {
    const errorString = compilerErrorsToString(inputs, errors);
    throw new Error(`Failed to compile:\n${errorString}`);
  }
  return compose(migrate(ast));
}

/**
 * Creates dummy query executor wich always return empty row.
 *
 * TODO: add abbility to define dummy return data
 */
export function mockQueryExecutor(): QueryExecutor {
  return {
    executeQuery(
      _def: Definition,
      _q: QueryDef,
      _ctx: Storage,
      _contextIds: number[]
    ): Promise<NestedRow[]> {
      return Promise.resolve([]);
    },

    executeQueryTree(
      _def: Definition,
      _qt: QueryTree,
      _ctx: Storage,
      _contextIds: number[]
    ): Promise<NestedRow[]> {
      return Promise.resolve([]);
    },
  };
}
/**
 * Provide models blueprint, and query blueprint separately.
 * This function will compile a blueprint and extract the query definition,
 * so you can use it in tests.
 */

export function makeTestQuery(models: string, query: string): { def: Definition; query: QueryDef } {
  const bp = `
  ${models}

  model TestHelperModel {}
  api {
    entrypoint TestHelperModel {
      custom endpoint {
        method GET
        cardinality many
        path "/test"
        action {
          ${query}
        }
      }
    }
  }
  `;
  const def = compileFromString(bp);
  const endpoint = def.apis[0].entrypoints[0].endpoints[0] as CustomManyEndpointDef;
  const action = endpoint.actions[0] as QueryAction;
  return { def, query: action.query };
}
