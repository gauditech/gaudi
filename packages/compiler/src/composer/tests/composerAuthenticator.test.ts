import { compileFromString } from "@compiler/common/testUtils";

describe("authenticator composer", () => {
  it("succeeds for simple authenticator", () => {
    const bp = `
      model AuthUser {}
      auth { model AuthUser }
    `;

    const def = compileFromString(bp);

    expect(def.authenticator).toMatchSnapshot();
  });

  it("@auth resolves to auth's model", () => {
    const bp = `
      model AuthUser {}
      auth { model AuthUser }
      api {
        entrypoint AuthUser {
          // @auth should resolve to auth's model
          authorize { @auth.id is not null}
          list endpoint {}
        }
      }
    `;

    const def = compileFromString(bp);

    expect(def.apis[0].entrypoints[0].endpoints[0].authorize).toMatchSnapshot();
  });
});
