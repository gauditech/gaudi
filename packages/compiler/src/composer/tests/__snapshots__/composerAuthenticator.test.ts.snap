// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticator composer resolves authenticator model implicit relations 1`] = `
{
  "fromModel": "UserProfile",
  "fromModelRefKey": "UserProfile",
  "kind": "relation",
  "modelRefKey": "AuthUser",
  "name": "userProfileUserRel",
  "refKey": "AuthUser.userProfileUserRel",
  "through": "user",
  "throughRefKey": "UserProfile.user",
  "unique": false,
}
`;

exports[`authenticator composer succeeds for simple authenticator 1`] = `
{
  "accessTokenModel": {
    "name": "AuthUserAccessToken",
    "refKey": "AuthUserAccessToken",
  },
  "authUserModel": {
    "name": "AuthUser",
    "refKey": "AuthUser",
  },
  "method": {
    "kind": "basic",
  },
  "name": "Auth",
}
`;

exports[`authenticator composer succeeds for simple authenticator 2`] = `
[
  {
    "aggregates": [],
    "computeds": [],
    "dbname": "authuser",
    "fields": [
      {
        "dbname": "id",
        "kind": "field",
        "modelRefKey": "AuthUser",
        "name": "id",
        "nullable": false,
        "primary": true,
        "refKey": "AuthUser.id",
        "type": "integer",
        "unique": true,
        "validate": undefined,
      },
      {
        "dbname": "name",
        "kind": "field",
        "modelRefKey": "AuthUser",
        "name": "name",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUser.name",
        "type": "string",
        "unique": false,
        "validate": {
          "args": [
            {
              "kind": "literal",
              "literal": {
                "kind": "integer",
                "value": 1,
              },
            },
          ],
          "kind": "call",
          "validator": "minLength",
        },
      },
      {
        "dbname": "username",
        "kind": "field",
        "modelRefKey": "AuthUser",
        "name": "username",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUser.username",
        "type": "string",
        "unique": true,
        "validate": {
          "args": [
            {
              "kind": "literal",
              "literal": {
                "kind": "integer",
                "value": 8,
              },
            },
          ],
          "kind": "call",
          "validator": "minLength",
        },
      },
      {
        "dbname": "passwordhash",
        "kind": "field",
        "modelRefKey": "AuthUser",
        "name": "passwordHash",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUser.passwordHash",
        "type": "string",
        "unique": false,
        "validate": {
          "args": [
            {
              "kind": "literal",
              "literal": {
                "kind": "integer",
                "value": 8,
              },
            },
          ],
          "kind": "call",
          "validator": "minLength",
        },
      },
    ],
    "hooks": [],
    "kind": "model",
    "name": "AuthUser",
    "queries": [],
    "refKey": "AuthUser",
    "references": [],
    "relations": [
      {
        "fromModel": "AuthUserAccessToken",
        "fromModelRefKey": "AuthUserAccessToken",
        "kind": "relation",
        "modelRefKey": "AuthUser",
        "name": "tokens",
        "refKey": "AuthUser.tokens",
        "through": "authUser",
        "throughRefKey": "AuthUserAccessToken.authUser",
        "unique": false,
      },
    ],
  },
  {
    "aggregates": [],
    "computeds": [],
    "dbname": "authuseraccesstoken",
    "fields": [
      {
        "dbname": "id",
        "kind": "field",
        "modelRefKey": "AuthUserAccessToken",
        "name": "id",
        "nullable": false,
        "primary": true,
        "refKey": "AuthUserAccessToken.id",
        "type": "integer",
        "unique": true,
        "validate": undefined,
      },
      {
        "dbname": "token",
        "kind": "field",
        "modelRefKey": "AuthUserAccessToken",
        "name": "token",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUserAccessToken.token",
        "type": "string",
        "unique": true,
        "validate": undefined,
      },
      {
        "dbname": "expirydate",
        "kind": "field",
        "modelRefKey": "AuthUserAccessToken",
        "name": "expiryDate",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUserAccessToken.expiryDate",
        "type": "string",
        "unique": false,
        "validate": undefined,
      },
      {
        "dbname": "authuser_id",
        "kind": "field",
        "modelRefKey": "AuthUserAccessToken",
        "name": "authUser_id",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUserAccessToken.authUser_id",
        "type": "integer",
        "unique": false,
        "validate": undefined,
      },
    ],
    "hooks": [],
    "kind": "model",
    "name": "AuthUserAccessToken",
    "queries": [],
    "refKey": "AuthUserAccessToken",
    "references": [
      {
        "fieldRefKey": "AuthUserAccessToken.authUser_id",
        "kind": "reference",
        "modelRefKey": "AuthUserAccessToken",
        "name": "authUser",
        "nullable": false,
        "onDelete": undefined,
        "refKey": "AuthUserAccessToken.authUser",
        "toModelFieldRefKey": "AuthUser.id",
        "toModelRefKey": "AuthUser",
        "unique": false,
      },
    ],
    "relations": [],
  },
]
`;

exports[`authenticator composer succeeds for simple authenticator 3`] = `
[
  {
    "endpoints": [
      {
        "actions": [
          {
            "alias": "existingAuthUser",
            "kind": "query",
            "model": "AuthUser",
            "query": {
              "filter": {
                "args": [
                  {
                    "kind": "identifier-path",
                    "namePath": [
                      "AuthUser",
                      "username",
                    ],
                  },
                  {
                    "kind": "alias-reference",
                    "path": [
                      "username",
                    ],
                    "source": "fieldset",
                  },
                ],
                "kind": "function",
                "name": "is",
              },
              "fromPath": [
                "AuthUser",
              ],
              "kind": "query",
              "limit": 1,
              "modelRefKey": "AuthUser",
              "name": "$query",
              "offset": undefined,
              "orderBy": undefined,
              "refKey": "N/A",
              "retCardinality": "one",
              "retType": "AuthUser",
              "select": [
                {
                  "alias": "passwordHash",
                  "expr": {
                    "kind": "identifier-path",
                    "namePath": [
                      "AuthUser",
                      "passwordHash",
                    ],
                  },
                  "kind": "expression",
                  "type": {
                    "kind": "string",
                    "nullable": false,
                  },
                },
                {
                  "alias": "id",
                  "expr": {
                    "kind": "identifier-path",
                    "namePath": [
                      "AuthUser",
                      "id",
                    ],
                  },
                  "kind": "expression",
                  "type": {
                    "kind": "integer",
                    "nullable": false,
                  },
                },
              ],
            },
          },
          {
            "alias": "$action_1",
            "hook": {
              "args": [
                {
                  "kind": "basic",
                  "name": "clearPassword",
                  "setter": {
                    "kind": "alias-reference",
                    "path": [
                      "password",
                    ],
                    "source": "fieldset",
                  },
                },
                {
                  "kind": "basic",
                  "name": "hashPassword",
                  "setter": {
                    "kind": "alias-reference",
                    "path": [
                      "existingAuthUser",
                      "passwordHash",
                    ],
                    "source": "aliases",
                  },
                },
              ],
              "hook": {
                "file": "hooks/actions.js",
                "kind": "source",
                "runtimeName": "@GAUDI_INTERNAL",
                "target": "authenticateUser",
              },
            },
            "kind": "execute-hook",
            "responds": false,
          },
          {
            "alias": "accessToken",
            "changeset": [
              {
                "kind": "basic",
                "name": "token",
                "setter": {
                  "args": [
                    {
                      "kind": "literal",
                      "literal": {
                        "kind": "integer",
                        "value": 32,
                      },
                    },
                  ],
                  "kind": "function",
                  "name": "cryptoToken",
                },
              },
              {
                "kind": "basic",
                "name": "expiryDate",
                "setter": {
                  "args": [
                    {
                      "args": [
                        {
                          "args": [],
                          "kind": "function",
                          "name": "now",
                        },
                        {
                          "kind": "literal",
                          "literal": {
                            "kind": "integer",
                            "value": 3600000,
                          },
                        },
                      ],
                      "kind": "function",
                      "name": "+",
                    },
                  ],
                  "kind": "function",
                  "name": "stringify",
                },
              },
              {
                "kind": "basic",
                "name": "authUser_id",
                "setter": {
                  "kind": "alias-reference",
                  "path": [
                    "existingAuthUser",
                    "id",
                  ],
                  "source": "aliases",
                },
              },
            ],
            "isPrimary": false,
            "kind": "create-one",
            "model": "AuthUserAccessToken",
            "select": [
              {
                "alias": "token",
                "expr": {
                  "kind": "identifier-path",
                  "namePath": [
                    "accessToken",
                    "token",
                  ],
                },
                "kind": "expression",
                "type": {
                  "kind": "string",
                  "nullable": false,
                },
              },
              {
                "alias": "id",
                "expr": {
                  "kind": "identifier-path",
                  "namePath": [
                    "accessToken",
                    "id",
                  ],
                },
                "kind": "expression",
                "type": {
                  "kind": "integer",
                  "nullable": false,
                },
              },
            ],
            "targetPath": [
              "AuthUserAccessToken",
            ],
          },
          {
            "alias": "$action_3",
            "hook": {
              "args": [
                {
                  "kind": "basic",
                  "name": "status",
                  "setter": {
                    "kind": "literal",
                    "literal": {
                      "kind": "integer",
                      "value": 200,
                    },
                  },
                },
                {
                  "kind": "basic",
                  "name": "body_token",
                  "setter": {
                    "kind": "alias-reference",
                    "path": [
                      "accessToken",
                      "token",
                    ],
                    "source": "aliases",
                  },
                },
              ],
              "hook": {
                "file": "hooks/actions.js",
                "kind": "source",
                "runtimeName": "@GAUDI_INTERNAL",
                "target": "sendResponse",
              },
            },
            "kind": "execute-hook",
            "responds": true,
          },
        ],
        "authSelect": [],
        "authorize": undefined,
        "fieldset": {
          "kind": "record",
          "nullable": false,
          "record": {
            "password": {
              "kind": "field",
              "nullable": false,
              "required": true,
              "type": "string",
              "validate": undefined,
            },
            "username": {
              "kind": "field",
              "nullable": false,
              "required": true,
              "type": "string",
              "validate": undefined,
            },
          },
        },
        "kind": "custom-many",
        "method": "POST",
        "parentContext": [],
        "path": "login",
        "responds": false,
        "response": undefined,
        "target": {
          "alias": "$target_0",
          "kind": "model",
          "name": "AuthUser",
          "namePath": [
            "AuthUser",
          ],
          "retType": "AuthUser",
          "select": [
            {
              "alias": "id",
              "expr": {
                "kind": "identifier-path",
                "namePath": [
                  "AuthUser",
                  "id",
                ],
              },
              "kind": "expression",
              "type": {
                "kind": "integer",
                "nullable": false,
              },
            },
          ],
        },
      },
      {
        "actions": [
          {
            "alias": "accessToken",
            "kind": "query",
            "model": "AuthUserAccessToken",
            "query": {
              "filter": {
                "args": [
                  {
                    "kind": "identifier-path",
                    "namePath": [
                      "AuthUserAccessToken",
                      "token",
                    ],
                  },
                  {
                    "kind": "alias-reference",
                    "path": [
                      "@requestAuthToken",
                    ],
                    "source": "aliases",
                  },
                ],
                "kind": "function",
                "name": "is",
              },
              "fromPath": [
                "AuthUserAccessToken",
              ],
              "kind": "query",
              "limit": 1,
              "modelRefKey": "AuthUserAccessToken",
              "name": "$query",
              "offset": undefined,
              "orderBy": undefined,
              "refKey": "N/A",
              "retCardinality": "one",
              "retType": "AuthUserAccessToken",
              "select": [
                {
                  "alias": "id",
                  "expr": {
                    "kind": "identifier-path",
                    "namePath": [
                      "AuthUserAccessToken",
                      "id",
                    ],
                  },
                  "kind": "expression",
                  "type": {
                    "kind": "integer",
                    "nullable": false,
                  },
                },
              ],
            },
          },
          {
            "kind": "delete-one",
            "model": "AuthUserAccessToken",
            "targetPath": [
              "accessToken",
            ],
          },
          {
            "alias": "$action_2",
            "hook": {
              "args": [
                {
                  "kind": "basic",
                  "name": "status",
                  "setter": {
                    "kind": "literal",
                    "literal": {
                      "kind": "integer",
                      "value": 204,
                    },
                  },
                },
              ],
              "hook": {
                "file": "hooks/actions.js",
                "kind": "source",
                "runtimeName": "@GAUDI_INTERNAL",
                "target": "sendResponse",
              },
            },
            "kind": "execute-hook",
            "responds": true,
          },
        ],
        "authSelect": [],
        "authorize": undefined,
        "fieldset": undefined,
        "kind": "custom-many",
        "method": "POST",
        "parentContext": [],
        "path": "logout",
        "responds": false,
        "response": undefined,
        "target": {
          "alias": "$target_0",
          "kind": "model",
          "name": "AuthUser",
          "namePath": [
            "AuthUser",
          ],
          "retType": "AuthUser",
          "select": [
            {
              "alias": "id",
              "expr": {
                "kind": "identifier-path",
                "namePath": [
                  "AuthUser",
                  "id",
                ],
              },
              "kind": "expression",
              "type": {
                "kind": "integer",
                "nullable": false,
              },
            },
          ],
        },
      },
      {
        "actions": [
          {
            "alias": "authUser",
            "changeset": [
              {
                "fieldsetPath": [
                  "authUser",
                  "name",
                ],
                "kind": "input",
                "name": "name",
                "setter": {
                  "kind": "alias-reference",
                  "path": [
                    "authUser",
                    "name",
                  ],
                  "source": "fieldset",
                },
                "validate": undefined,
              },
              {
                "fieldsetPath": [
                  "authUser",
                  "username",
                ],
                "kind": "input",
                "name": "username",
                "setter": {
                  "kind": "alias-reference",
                  "path": [
                    "authUser",
                    "username",
                  ],
                  "source": "fieldset",
                },
                "validate": undefined,
              },
              {
                "kind": "basic",
                "name": "passwordHash",
                "setter": {
                  "args": [
                    {
                      "kind": "alias-reference",
                      "path": [
                        "password",
                      ],
                      "source": "fieldset",
                    },
                    {
                      "kind": "literal",
                      "literal": {
                        "kind": "integer",
                        "value": 10,
                      },
                    },
                  ],
                  "kind": "function",
                  "name": "cryptoHash",
                },
              },
            ],
            "isPrimary": false,
            "kind": "create-one",
            "model": "AuthUser",
            "select": [
              {
                "alias": "id",
                "expr": {
                  "kind": "identifier-path",
                  "namePath": [
                    "authUser",
                    "id",
                  ],
                },
                "kind": "expression",
                "type": {
                  "kind": "integer",
                  "nullable": false,
                },
              },
              {
                "alias": "name",
                "expr": {
                  "kind": "identifier-path",
                  "namePath": [
                    "authUser",
                    "name",
                  ],
                },
                "kind": "expression",
                "type": {
                  "kind": "string",
                  "nullable": false,
                },
              },
              {
                "alias": "username",
                "expr": {
                  "kind": "identifier-path",
                  "namePath": [
                    "authUser",
                    "username",
                  ],
                },
                "kind": "expression",
                "type": {
                  "kind": "string",
                  "nullable": false,
                },
              },
            ],
            "targetPath": [
              "AuthUser",
            ],
          },
          {
            "alias": "$action_1",
            "hook": {
              "args": [
                {
                  "kind": "basic",
                  "name": "status",
                  "setter": {
                    "kind": "literal",
                    "literal": {
                      "kind": "integer",
                      "value": 201,
                    },
                  },
                },
                {
                  "kind": "basic",
                  "name": "body_id",
                  "setter": {
                    "kind": "alias-reference",
                    "path": [
                      "authUser",
                      "id",
                    ],
                    "source": "aliases",
                  },
                },
                {
                  "kind": "basic",
                  "name": "body_name",
                  "setter": {
                    "kind": "alias-reference",
                    "path": [
                      "authUser",
                      "name",
                    ],
                    "source": "aliases",
                  },
                },
                {
                  "kind": "basic",
                  "name": "body_username",
                  "setter": {
                    "kind": "alias-reference",
                    "path": [
                      "authUser",
                      "username",
                    ],
                    "source": "aliases",
                  },
                },
              ],
              "hook": {
                "file": "hooks/actions.js",
                "kind": "source",
                "runtimeName": "@GAUDI_INTERNAL",
                "target": "sendResponse",
              },
            },
            "kind": "execute-hook",
            "responds": true,
          },
        ],
        "authSelect": [],
        "authorize": undefined,
        "fieldset": {
          "kind": "record",
          "nullable": false,
          "record": {
            "authUser": {
              "kind": "record",
              "nullable": false,
              "record": {
                "name": {
                  "kind": "field",
                  "nullable": false,
                  "required": true,
                  "type": "string",
                  "validate": {
                    "args": [
                      {
                        "kind": "literal",
                        "literal": {
                          "kind": "integer",
                          "value": 1,
                        },
                      },
                    ],
                    "kind": "call",
                    "validator": "minLength",
                  },
                },
                "username": {
                  "kind": "field",
                  "nullable": false,
                  "required": true,
                  "type": "string",
                  "validate": {
                    "args": [
                      {
                        "kind": "literal",
                        "literal": {
                          "kind": "integer",
                          "value": 8,
                        },
                      },
                    ],
                    "kind": "call",
                    "validator": "minLength",
                  },
                },
              },
            },
            "password": {
              "kind": "field",
              "nullable": false,
              "required": true,
              "type": "string",
              "validate": {
                "args": [
                  {
                    "kind": "literal",
                    "literal": {
                      "kind": "integer",
                      "value": 8,
                    },
                  },
                ],
                "kind": "call",
                "validator": "minLength",
              },
            },
          },
        },
        "kind": "custom-many",
        "method": "POST",
        "parentContext": [],
        "path": "register",
        "responds": false,
        "response": undefined,
        "target": {
          "alias": "$target_0",
          "kind": "model",
          "name": "AuthUser",
          "namePath": [
            "AuthUser",
          ],
          "retType": "AuthUser",
          "select": [
            {
              "alias": "id",
              "expr": {
                "kind": "identifier-path",
                "namePath": [
                  "AuthUser",
                  "id",
                ],
              },
              "kind": "expression",
              "type": {
                "kind": "integer",
                "nullable": false,
              },
            },
          ],
        },
      },
    ],
    "entrypoints": [],
    "name": "",
    "target": {
      "alias": "$target_0",
      "identifyWith": {
        "paramName": "authuser_id",
        "path": [
          "id",
        ],
        "type": "integer",
      },
      "kind": "model",
      "name": "AuthUser",
      "namePath": [
        "AuthUser",
      ],
      "retType": "AuthUser",
    },
  },
]
`;
